File: 2025-10-October-GFG-POTD/October 08 - Construct Tree from Preorder & Postorder/Solution.java
/*
class Node {
    int data;
    Node left, right;

    Node(int val) {
        data = val;
        left = right = null;
    }
}
*/

class Solution {
     int preIndex = 0;
    Map<Integer, Integer> postMap = new HashMap<>();

    public Node constructTree(int[] pre, int[] post) {
        // Map to store value -> index in postorder for O(1) lookups
        for (int i = 0; i < post.length; i++) {
            postMap.put(post[i], i);
        }
        return build(pre, post, 0, post.length - 1);
    }

    private Node build(int[] pre, int[] post, int l, int r) {
        if (preIndex >= pre.length || l > r) return null;

        // Create root from preorder
        Node root = new Node(pre[preIndex++]);

        // If only one node, return it
        if (l == r || preIndex >= pre.length) return root;

        // Find index of next preorder element in postorder
        int postIndex = postMap.get(pre[preIndex]);

        // Build left and right subtrees
        root.left = build(pre, post, l, postIndex);
        root.right = build(pre, post, postIndex + 1, r - 1);

        return root;
    }
}

File: 2025-10-October-GFG-POTD/October 08 - Construct Tree from Preorder & Postorder/Solution.py
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        self.preIndex = 0
        self.postMap = {}

    def constructTree(self, pre, post):
        # Map value -> index in postorder
        for i, val in enumerate(post):
            self.postMap[val] = i
        return self._build(pre, post, 0, len(post) - 1)

    def _build(self, pre, post, l, r):
        if self.preIndex >= len(pre) or l > r:
            return None

        root = Node(pre[self.preIndex])
        self.preIndex += 1

        if l == r or self.preIndex >= len(pre):
            return root

        postIndex = self.postMap[pre[self.preIndex]]

        root.left = self._build(pre, post, l, postIndex)
        root.right = self._build(pre, post, postIndex + 1, r - 1)

        return root
